message_app/send_message.py

```diff
@@ -160,10 +159,27 @@ def message_handler(data):
     emit("message_handler_client", return_data, room=conversation_id)
 
 # A socket that updates user's last_active_time
-@socketio.on("last_active", namespace="/messages")
-def last_active(data):
+@socketio.on("last_active_time", namespace="/messages")
+def last_active_time(data):
+    # The data is a dictionary with three keys:
+    # username: username
+    # last_active_time: new last_active_time to update in the database
+    # is_closing: a boolean key that track if the user close their browser
     username = data["username"]
     last_active_time = data["last_active_time"]
+    is_closing = data["is_closing"]
     current_user = User.select(username)
+    old_status = check_user_status(current_user.last_active_time)
+    new_status = "away" if is_closing else check_user_status(last_active_time)
     current_user.last_active_time = last_active_time
     DB.session.commit()
+
+    if (old_status != new_status):
+        return_data = {
+            "username": username,
+            "status": new_status
+        }
+        emit("update_conversations_container", return_data, broadcast=True, include_self=False)
+
+def check_user_status(last_active_time: int) -> str:
+    return "active" if (time.time() * 1000 - last_active_time < LAST_ACTIVE_INTERVAL) else "away"
```

Comment:
The naming here is not good
- Last active time of user: what user, the current logged in user, or their friends?
- `is_closing`? So it is being closed, but not closed yet?

Reply:
Yes, it is being closed, but not completely closed yet. I call the function inside the [beforeunload event](https://developer.mozilla.org/en-US/docs/Web/API/Window/beforeunload_event), so technically the browser is being closed at the time this event happened, but not completely closed yet.